projecttacoma/measure-repository-service
# ZOD (request param verification)

# Summary
Added the ZOD dependency and implemented request input schemas for each of the defined operations

## New behavior
All existing request param input validation should continue to work, but now invalid params will throw 400 errors, unsupported params will throw 501 errors, and malformed params (url not a valid url) will throw 400 errors

## Code changes
 - Created request schemas for `search`, `$package`, and `$data-requirements` operations. Set all expected schema values to string, except for url for the time being.
 - Created dynamic functions to validate input and return proper OperationOutcome responses
 - Updated LibraryService and MeasureService files to use new requestSchemas
 - renamed `validationUtils` to `inputUtils` since all validation is now handled by Zod
 - Updated and added testing
 - Updated `.gitignore` to ignore eqcm-content-r4-2021 repo

# Testing guidance
 - `npm run db:reset`
 - `npm run db:loadBundle "ecqm-content-r4-2021/bundles/measure/ColorectalCancerScreeningsFHIR/ColorectalCancerScreeningsFHIR-bundle.json"`
 - `npm start`
 - Try all the requests in the included insomnia bundle and make sure all outcomes make sense
 - Check the schemas against the spec [here](https://build.fhir.org/ig/HL7/cqf-measures/measure-repository-service.html) and make sure all params are accounted for.
 - NOTE: while `periodStart` and `periodEnd` are not currently listed as valid params for `$data-requirements` operations, we believe this is a mistake, and so they are included in the schemas

[ZOD_testing.json.zip](https://github.com/projecttacoma/measure-repository-service/files/10676504/ZOD_testing.json.zip)
